### PR TITLE
FormTokenField: Hide suggestions list on blur event if input value is invalid

### PR DESCRIPTION
**Pull request by shvlv**
## What?
Currently, if FormTokenField loses focus the suggestions list is still rendered unless the input value is empty. The PR forces input clearing and suggestion list closing on blur event even if the input is not empty but invalid based on the `__experimentalValidateInput` property.

## Why?
In my opinion, the change allows for improving UX because it is not clear why the need suggestions list is still opened after blurring (i.e. the user finished interaction) if there is no valid information.

## How?
`inputHasValidValue` is a bit misleading now because we have `__experimentalValidateInput`. Basically, the first one uses `saveTransform` under the hood which is useful during suggestion searching and matching, and saving, the second one validates the token immediately before persisting.
IMO `inputHasValidValue() && __experimentalValidateInput( incompleteTokenValue )` is the right condition to detect if the input is "valid" (makes sense) after blurring.

## Testing Instructions
Changes could be reproduced if the FormTokenField component uses `__experimentalValidateInput` property. In this case, if the input value doesn't pass the validation suggestion list gets hidden after the field loses focus.

### Testing Instructions for Keyboard
The change works if you switch to the next active element with the `Tab` button.

### Before
[before.webm](https://user-images.githubusercontent.com/3157352/224801768-88f23a10-8637-497a-95ed-4ffee2ffec6f.webm)

### After
[after.webm](https://user-images.githubusercontent.com/3157352/224801795-56a1e363-2e5b-4051-b4cc-cb0ca685bd05.webm)

### CODE REVIEW

**Comment by ciampo** [comment]
Thank you for working on this, @shvlv ! I appreciate the detailed PR description and the addition of unit tests!

Before continuing in the review process, I'd love to hear opinions from design folks too, @jasmussen @jameskoster @SaxonF

**Comment by jameskoster** [comment]
Thanks for the ping. 

This makes sense to me in principle but I wasn't able to test in Storybook because I cannot set `__experimentalValidateInput`.

**Review by chad1008 — COMMENTED** [review]
👍 This change makes sense to me, I agree that it's a nice UI improvement. The changes look good and test well for me - thanks for adding tests at the same time.

I tested in Storybook by locally adding a quick story to trial the prop:

```diff

diff --git a/packages/components/src/form-token-field/stories/index.tsx b/packages/components/src/form-token-field/stories/index.tsx
index d4af720913..51997ba17b 100644
--- a/packages/components/src/form-token-field/stories/index.tsx
+++ b/packages/components/src/form-token-field/stories/index.tsx
@@ -123,3 +123,11 @@ WithCustomRenderItem.args = {
 		<div>{ `${ item } — a nice place to visit` }</div>
 	),
 };
+
+export const WithValidatedInput: ComponentStory< typeof FormTokenField > =
+	DefaultTemplate.bind( {} );
+WithValidatedInput.args = {
+	...Default.args,
+	__experimentalValidateInput: ( input: string ) =>
+		continents.includes( input ),
+};

```

I'm not sure if we'd actually want to add a story like that for an experimental prop though...

I'll defer to @mirka or @ciampo on the actual approval, they have better eyes for our usual testing patterns than I do.

**Comment by ciampo** [comment]
Thank you @jameskoster and @chad1008 ! I'll go ahead and take a deeper look at the code changes.

@shvlv , in the meantime could you add a "before" and "after" screen recording of the component, so that more folks are going to be able to understand the result of the proposed changes?

File: packages/components/src/form-token-field/test/index.tsx
Comment (ciampo) [inline packages/components/src/form-token-field/test/index.tsx]: Noting that this test would pass also without the changes proposed by this PR.

Actually, the suggestion list should not be visible already after the input is cleared. Blurring the input in this case shouldn't make any differences.

Could we tweak this unit test by removing the `await user.click( document.body );` action and updating the test name to something like "should hide the suggestion list on an empty input" ?

**Review by ciampo — COMMENTED** [review]
Great work here, @shvlv !

Could you add an entry to the package's [CHANGELOG](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md) ?


<details>
<summary>
@chad1008 's suggestion makes sense to me. We could add a new storybook example that showcases how to use the `__experimentalValidateInput` props</summary>

```diff
diff --git a/packages/components/src/form-token-field/stories/index.tsx b/packages/components/src/form-token-field/stories/index.tsx
index d4af720913..0657e49917 100644
--- a/packages/components/src/form-token-field/stories/index.tsx
+++ b/packages/components/src/form-token-field/stories/index.tsx
@@ -123,3 +123,17 @@ WithCustomRenderItem.args = {
 		<div>{ `${ item } — a nice place to visit` }</div>
 	),
 };
+
+/**
+ * Only values for which the `__experimentalValidateInput` function returns
+ * `true` will be tokenized. (This is still an experimental feature and is
+ * subject to change.)
+ */
+export const WithCustomInputValidation: ComponentStory<
+	typeof FormTokenField
+> = DefaultTemplate.bind( {} );
+WithCustomInputValidation.args = {
+	...Default.args,
+	__experimentalValidateInput: ( input: string ) =>
+		continents.includes( input ),
+};

```
</details>

**Comment by shvlv** [comment]
Oh, it seems the wrong rebase. Sorry :disappointed:

**Comment by shvlv** [comment]
Hi, @ciampo I've applied your suggestions and updated the PR description with before/after small screencasts. Let me know if something else is needed.

**Review by ciampo — APPROVED** [review]
LGTM 🚀 

Before merging, would you mind also applying [my previous suggestion](https://github.com/WordPress/gutenberg/pull/48785#pullrequestreview-1334565004) about adding an additional Storybook example?

Thank you 🙏

**Comment by shvlv** [comment]
A storybook example was updated. It's a bit confusing to me because visual code representation doesn't contain the passed callback but it looks like the Storybook problem :man_shrugging: 
image

File: packages/components/src/form-token-field/stories/index.tsx
Comment (ciampo) [inline packages/components/src/form-token-field/stories/index.tsx:132]: Let's add an inline comment to the story too

```suggestion
/**
 * Only values for which the `__experimentalValidateInput` function returns
 * `true` will be tokenized. (This is still an experimental feature and is
 * subject to change.)
 */
export const WithValidatedInput: ComponentStory< typeof FormTokenField > =
```

**Review by ciampo — COMMENTED** [review]
> It's a bit confusing to me because visual code representation doesn't contain the passed callback but it looks like the Storybook problem 🤷‍♂️

That's correct — it is a known Storybook bug (I couldn't find the exact issue but I've read about it before).

**Comment by ciampo** [comment]
Thank you again for your contribution, @shvlv 🙏

**Comment by desrosj** [comment]
@shvlv If you have a moment, could you connect your GitHub account to your wordpress.org one? This change is slated to be included in the upcoming WordPress 6.3 release on August 8th, and it would be great if we could give you attribution by mentioning you on the credits screen. You can find more details on how contributions are tracked and how connecting GitHub accounts helps in [this post on the Make WordPress Core blog](https://make.wordpress.org/core/2020/03/19/associating-github-accounts-with-wordpress-org-profiles/).

**Comment by shvlv** [comment]
Hi @desrosj , I did it several months ago. I see the following in my wordpress.org [profile](https://profiles.wordpress.org/shvv/). Is it something wrong?
image

**Comment by desrosj** [comment]
Ah! Sorry about that. Looks like that connection was missed for you in our collection process. I'll update the list to reflect that. Thank you for your contribution!

**Pull request merged**
